Notifications Widget

monkins1010 opened the pull request.

michaeltout commented on src/components/SendModal/ProvisionIdentity/ProvisionIdentityConfirm/ProvisionIdentityConfirm.js:
```diff
@@ -103,31 +114,51 @@ class ProvisionIdentityConfirm extends Component {
+      if (confirmedProvisionedID.decision.result.fully_qualified_name.split('.')[0].toLowerCase() !== identityName.split('.')[0].toLowerCase()
+            || confirmedProvisionedID.decision.result.parent !== parent) throw new Error('Response does not match the expected ID request');
+
+      const notification = {
```
Notifications should be standardised, dispatched, edited, added, and removed using the functions provided in src/actions/actions/notifications/dispatchers/notifications.js, and the classes provided in src/utils/notification.js. This way, they can fit into the existing notification reducer, and be used for all sorts of non-verusid related purposes in the future. I would advise changing what you currently call the "SET_VERUSID_NOTIFICATIONS" action to a "SET_PENDING_VERUSIDS" action, and store specific information regarding the status, login request, info, etc. in there, and then when you need to show a notification on the home screen, or edit a notification to reflect a change, use dispatchAddNotification, or dispatchSetNotifications, and the relevant class for the desired notification type in src/utils/notification.js.

Different notification types display differently and can be used for different informtation, but there is no VerusID specific notification type.

michaeltout commented on src/actions/actions/services/dispatchers/UpdateServiceNotifications.js:
```diff
@@ -0,0 +1,26 @@
+    };
+  };
+  
+  export const updateServiceNotifications = (state, dispatch, channels) =>
```
Having an updateServiceNotifications function is good, but as mentioned before, it should follow the schema of check service data -> find out what needs to be notified -> push general notifications (not service specific) if anything should be notified.

michaeltout commented on src/containers/Home/HomeWidgets/NotificationWidget.js:
I believe this component needs to be refactored to be a general component with no VerusID specific aspects to it. Ideally, it should handle every notification type class in src/utils/notification.js, and display a different type of notification based on the data contained in each notification type. If you want to add icons or image uris to the notification types in src/utils/notification.js so that the notifications pushed from the VerusID store can have a logo, that's ok, we can discuss.

michaeltout commented on src/reducers/channelStores/verusid.js:
```diff
@@ -29,13 +32,19 @@ export const channelStore_verusid = (state = {
         ...state,
         watchedVerusIds: action.payload.watchedVerusIds
       }
+    case SET_VERUSID_NOTIFICATIONS:
```
As mentioned in another comment, this is the bit I advise to change to SET_PENDING_VERUSIDS

michaeltout commented on src/containers/DeepLink/LoginRequestComplete/LoginRequestComplete.js:
```diff
@@ -66,8 +66,13 @@ const LoginRequestComplete = props => {
+      const returnedData = await tryRedirect();
+      if (reDirect) {
+        // Only return to services, default is main home screen.
+        props.navigation.navigate("ServicesHome", { screen: reDirect, params: { data: returnedData.data }});
```
I don't think this should be called redirect for a few reasons. Firstly, a redirect for a deeplink is already a defined term (redirect back to website), and secondly, it forces the user to navigate back to ServicesHome. I think you should change this to "fromService" or "sourceServiceId" or something along those lines to better indicate what it is.

Also, if this loginConsentRequest contains a redirect instead of a webhook, isn't this a no-op? Shouldn't the navigation happen first, then tryRedirect gets called?

michaeltout commented on src/containers/DeepLink/LoginRequestInfo/LoginRequestInfo.js:
```diff
@@ -72,6 +79,82 @@ const LoginRequestInfo = props => {
+            tempMethod = tempMethod | 1;
+            tempdata = { data: req.challenge.requested_access[i].toJson().data, title: "Agreement to accept" }
+          }
+          // if (req.challenge.requested_access[i].vdxfkey === primitives.ATTESTATION_READ_REQUEST.vdxfid) {
```
Please remove this comment or clarify its role for future maintaining.

michaeltout commented on src/components/SendModal/ProvisionIdentity/ProvisionIdentityForm/ProvisionIdentityForm.js:
```diff
@@ -96,15 +97,22 @@ const ProvisionIdentityForm = (props) => {
                 }
+                if (idKey.vdxfkey == primitives.ID_PARENT_VDXF_KEY.vdxfid &&
+                  idKey.data != sendModal.coinObj.system_id) {
+                   parentname = `.${identity.result.identity.name}.${sendModal.coinObj.display_ticker}@`
```
I believe you have to use the fullyqualifiedname of the parent or risk leaving out some system or parent IDs

michaeltout reviewed (changes requested):
Hi Chris, thanks for the PR! Overall a good start, well on the way to implementing provisioning/notifications, but there are some key points to address, more of which is highlighted in the code comments (also this bullet point summary is not all the feedback in the code comments so I'd recommend going off of those):

- Refactor updateServiceNotifications to a general schema involving checking service data, determining notification requirements, and pushing non-service-specific notifications.

- Address handling of different provisioning response scenarios: immediate success, error, and pending approval without name commitment.

- Standardize notification handling using existing functions in src/actions/actions/notifications/dispatchers/notifications.js and classes in src/utils/notification.js.

- Change what you currently define as the "SET_VERUSID_NOTIFICATIONS" action to a more specific "SET_PENDING_VERUSIDS" action, while generalising how you deal with notifications

- Utilize dispatchAddNotification and dispatchSetNotifications for notification modifications, fitting them into the existing notification reducer.

- Refactor the Notification component to be generic, handling the genertic types of notifications from src/utils/notification.js

- Some general build files that are included to be removed

Feel free to contact me for any questions/clarification.

Best,
Michael

michaeltout commented on android/app/src/main/res/drawable-hdpi/node_modules_reactnavigation_src_views_assets_backicon.png:
Please remove this file

michaeltout commented on android/app/src/main/res/drawable-mdpi/src_images_customicons_valubrand.png:
Please remove this file

michaeltout commented on src/containers/Home/HomeWidgets/NotificationWidget.js:
Thanks, this file followed the discussed architecture for notifications now, but I'm somewhat confused about how a pending ID is represented before it becomes actionable by the user (e.g. while it is loading), as I see only deeplink notifications being dispatched, and the loading notification state isn't handled here yet.

michaeltout reviewed (changes requested):
Hi Chris, the changes are good, and I was able to fit in a quick review today (will take another more detailed look tomorrow), so I added comments into the code, and aside from some style comments, had a few functionality notes/questions (again, more detail in code):

1. It seems we're still waiting for provisioned IDs marked as completed by making the user sit while the tx confirms. It would be ideal if these were handled with notifications similarly to how the pending IDs are.

2. I'm a little confused about how an ID is represented before it is ready for the user to take action on, as it seems only deepink notifications are being dispatched.

Feel free to contact me for any more verification/in depth discussion, and thanks for the prompt changes.

\- Michael

monkins1010 commented:
> Hi Chris, the changes are good, and I was able to fit in a quick review today (will take another more detailed look tomorrow), so I added comments into the code, and aside from some style comments, had a few functionality notes/questions (again, more detail in code):
> 
> 1. It seems we're still waiting for provisioned IDs marked as completed by making the user sit while the tx confirms. It would be ideal if these were handled with notifications similarly to how the pending IDs are.
> 2. I'm a little confused about how an ID is represented before it is ready for the user to take action on, as it seems only deepink notifications are being dispatched.
> 
> Feel free to contact me for any more verification/in depth discussion, and thanks for the prompt changes.
> 
> - Michael



> Hi Chris, the changes are good, and I was able to fit in a quick review today (will take another more detailed look tomorrow), so I added comments into the code, and aside from some style comments, had a few functionality notes/questions (again, more detail in code):
> 
> 1. It seems we're still waiting for provisioned IDs marked as completed by making the user sit while the tx confirms. It would be ideal if these were handled with notifications similarly to how the pending IDs are.
> 2. I'm a little confused about how an ID is represented before it is ready for the user to take action on, as it seems only deepink notifications are being dispatched.
> 
> Feel free to contact me for any more verification/in depth discussion, and thanks for the prompt changes.
> 
> - Michael

C1) ID's marked with pending will exit out straightaway and be handled by the notifications. Yes ID's marked as complete await the transfer, and yes this could be up to 10 mins on a slow block so Ill make that exit out back to home screen so its not
C2) Yes a pending ID only exists in the background as data, not sure if the `Joe123@ Is being created` notification is necessary? if so can implement.

michaeltout commented on src/actions/actions/services/dispatchers/verusid/verusid.js:
```diff
@@ -56,3 +65,193 @@ export const unlinkVerusId = async (iAddress, chain) => {
+        let errorFound = false;
+        try {
+          if (pendingIds[ticker][iaddress].infoUri) {
+            const response = await axios.post(pendingIds[ticker][iaddress].infoUri);
```
This should be set to axios.get

michaeltout commented on src/actions/actions/services/dispatchers/verusid/verusid.js:
```diff
@@ -56,3 +65,193 @@ export const unlinkVerusId = async (iAddress, chain) => {
+          if (pendingIds[ticker][iaddress].infoUri) {
+            const response = await axios.post(pendingIds[ticker][iaddress].infoUri);
+
+            if (response.data.result.state === primitives.LOGIN_CONSENT_PROVISIONING_RESULT_STATE_FAILED.vdxfid) {
```
response.data should not be looked at here without it being wrapped in a `const responseData = new LoginConsentProvisioningResponse(response);` as to ensure that the response from the server is formatted correctly

michaeltout commented on src/components/SendModal/ProvisionIdentity/ProvisionIdentityConfirm/ProvisionIdentityConfirm.js:
```diff
@@ -103,31 +107,33 @@ class ProvisionIdentityConfirm extends Component {
+        this.props.sendModal.data.fromService, provisioningName, newLoadingNotification.uid, async (fqn) => {
+          
+          newLoadingNotification.body = "";
+          newLoadingNotification.title =  [`${fqn.split(".")[0]}@`, ` is being provisioned by `, `${provisioningName}@`]
```
This should use fqn for the provisioningName

michaeltout commented on src/actions/actions/services/dispatchers/verusid/verusid.js:
```diff
@@ -56,3 +65,193 @@ export const unlinkVerusId = async (iAddress, chain) => {
+        pendingIds[ticker][iaddress].status = NOTIFICATION_TYPE_VERUSID_READY;
+
+        await setRequestedVerusId(iaddress, pendingIds[ticker][iaddress], ticker);
+        await linkVerusId(iaddress, identity.result.fullyqualifiedname, ticker);
```
After this is linked, the user should be made aware if the linked ID has a different primary address included in it than the one in their wallet, and/or if the revocation and recovery IDs are not set to itself.

michaeltout commented on src/containers/Home/HomeWidgets/NotificationWidget.js:
```diff
@@ -0,0 +1,269 @@
+                            <TouchableOpacity
+                                onPress={() => { dispatchClearNotifications(); }}>
+                                {!collapsed && <Paragraph style={{ fontSize: 12, textDecorationLine: 'underline', marginVertical: 17 }}>
+                                    clear all
```
"Clear all" should have an "are you sure message", as well as clearing individual notifs. Later this maybe should be optional based on notif type or data but for now I think its ok to add an "are you sure" more generally to all.

michaeltout commented on src/actions/actions/services/dispatchers/verusid/verusid.js:
```diff
@@ -56,3 +65,193 @@ export const unlinkVerusId = async (iAddress, chain) => {
+              await updatePendingVerusIds();
+              newDeepLinkNotification.icon = NOTIFICATION_ICON_ERROR;
+              newDeepLinkNotification.uid = pendingIds[ticker][iaddress].notificationUid;
+              await dispatchAddNotification(newDeepLinkNotification);
```
dispatchAddNotification doesn't return a promise so doesn't need an `await`

michaeltout commented on src/actions/actions/services/dispatchers/UpdateServiceNotifications.js:
```diff
@@ -0,0 +1,26 @@
+      state,
+      dispatch,
+      channels,
+      SET_SERVICE_ACCOUNT,
```
This should work such that instead of dispatching SET_SERVICE_ACCOUNT, it dispatches a SET_SERVICE_NOTIFICATIONS action that follows the structure of all the other updateService<x> functions (action with {type: TYPE, payload: {channel: channel_name, body: {}}}) by making the updateServiceDataValue return {channel: channel_name, body: {}}. Its fine if the body is empty, as the function ultimately serves to dispatch notifications, but it should be a trackable redux event with a type. Also this means creating a ERROR_SERVICE_NOTIFICATIONS type and handling that in the error reducer so that errors with the notification function get stored in the redux store as well.

michaeltout commented on src/components/SendModal/ProvisionIdentity/ProvisionIdentityForm/ProvisionIdentityForm.js:
```diff
@@ -125,7 +134,8 @@ const ProvisionIdentityForm = (props) => {
       fromBase58Check(identity);
     } catch (e) {
-      if (!identity.endsWith('@')) {
+      const formattedId = state.parentname ? `${identity}${state.parentname}` : identity;
```
If the user submits an I-addr (which should be accepted in cases where the ID being provisioned already exists), the method of adding the parent fqn to the end of the data entered in the ID field will turn the i-address into an ID with fully qualified name of <i-address>.<parent>@. To solve this, I would recommend changing the form to throw an error if an I address is entered and the ID is going to be created with a parent, and I would change the header of the text box to no longer suggest i-addrs are allowed if a parent is going to be appended.

michaeltout commented on src/utils/notification.js:
```diff
@@ -74,18 +75,30 @@ export class LoadingNotification extends Notification {
 
 export class DeeplinkNotification extends Notification {
-  constructor(body, title, reopen = () => {}, uid, uri, acchash) {
+  constructor(body, title, reopen = () => {}, uid, uri, acchash, extraParams = null) {
```
The extraParams property is general (not easy to see what goes here) and seems unnecessary, please extrapolate the properties within it that seem to be put there for the Login request (iaddr, fqn, chain) from the uri in the deeplinknotification, and make fromService a property of deeplink notification. Alternatively, I might be missing something so we can discuss a different solution.

michaeltout commented on src/actions/actions/services/dispatchers/verusid/verusid.js:
```diff
@@ -56,3 +69,212 @@ export const unlinkVerusId = async (iAddress, chain) => {
+        try {
+          if (pendingIds[ticker][iaddress].infoUri) {
+            const response = await axios.get(pendingIds[ticker][iaddress].infoUri);
+            const responseData = new primitives.LoginConsentProvisioningDecision(response.data);
```
The server needs to return a signed LoginConsentProvisioningResponse, which the app then needs to verify, and ensure that the ID it is signed by matches the ID of the original loginconsentrequest. Refer to src/\_\_tests\_\_/vdxf/loginconsent.test.ts to see how the LoginConsentProvisioningResponse is used.

michaeltout commented on src/containers/Home/HomeWidgets/NotificationWidget.js:
```diff
@@ -0,0 +1,278 @@
+                        }}>
+                            <TouchableOpacity
+                                onPress={() => {
+                                    Alert.alert("Clear Notifications", "Are you sure you want to clear all notifications?",
```
If a notification for a pending VerusID is removed, the pending VerusID seems to stay in async storage forever, with no way to remove it. It would be ideal to either sweep pending VerusIDs on app start based on age (maybe older than a week or so) and then, when the notification is pressed, tell the user that the pending ID has expired if they try to link it (or automatically remove the notification, if it exists, once the pending ID is removed), or remove the pending ID when the notification is cleared. Another possible solution is to remove pending IDs that are not tied to any notification.

michaeltout commented on src/containers/Services/ServiceComponents/VerusIdService/VerusIdLogin.js:
```diff
@@ -0,0 +1,19 @@
+import { primitives } from 'verusid-ts-client'
+import { SET_DEEPLINK_DATA } from "../../../../utils/constants/storeType";
+
+export const processVerusId = async (props, loginConsentRequest, extraParams = null) => {
```
Please run files like these through a code formatter like prettier to avoid the spacing issues here, like the ending bracket being indented on a different level than the first line of the function (especially if its a newly created file so code formatting would not add new deltas).

michaeltout commented on src/reducers/channelStores/verusid.js:
```diff
@@ -29,13 +33,19 @@ export const channelStore_verusid = (state = {
     case SIGN_OUT_COMPLETE:
```
It seems that serviceChannelOpen stays true on SIGN_OUT_COMPLETE, which is inconsistent with how the previous wyre service implementation worked, and might possibly lead to calls being attempted from a state where the user isn't signed in?

michaeltout commented on src/containers/Services/ServiceComponents/VerusIdService/VerusIdLogin.js:
```diff
@@ -1,18 +1,18 @@
     const req = new primitives.LoginConsentRequest();
     req.fromBuffer(Buffer.from(loginConsentRequest, 'base64'));
-
+  console.log("req", req);
```
Please remove this console.log

michaeltout commented on src/containers/Home/HomeWidgets/NotificationWidget.js:
```diff
@@ -0,0 +1,283 @@
+
+        if (directory[uid].acchash === acchash) {
+            if (directory[uid].type === NOTIFICATION_TYPE_DEEPLINK) {
+                const tempDeepLinkNotification = DeeplinkNotification.fromJson(directory[uid], processVerusId);
```
Deeplink notifications are not verusid specific so we can't be calling processVerusId for all deeplink notifications as their onAction. There needs to be a way to process notification onActions and somehow link multiple different onAction functions to notifications here.

michaeltout commented on src/containers/DeepLink/LoginRequestIdentity/LoginRequestIdentity.js:
```diff
@@ -54,17 +93,72 @@ const LoginRequestIdentity = props => {
+    }
+  }, [extraParams])
+
+  useEffect(() => {
+
+    if (idsloaded && req.challenge.subject && req.challenge.subject.length > 0 &&
+      req.challenge.subject.some(item => item.vdxfkey === primitives.ID_ADDRESS_VDXF_KEY.vdxfid)) {
+
+        const providionedId = req.challenge.subject.find(item => item.vdxfkey === primitives.ID_ADDRESS_VDXF_KEY.vdxfid).data;
+        for (const chainId of activeCoinIds) { 
+          if (linkedIds[chainId] && Object.keys(linkedIds[chainId]).includes(providionedId)) {
+            return;
+          }
+        }
+      getIdentity(system_id, providionedId).then((provisionedID) => {
+        getPotentialPrimaryAddresses().then((addresses) => {
+          if (provisionedID.result) {
+            for (const address of provisionedID.result.identity.primaryaddresses) {
+              if (addresses.includes(address)) {
+                dispatch({
+                  type: SET_DEEPLINK_DATA_EXTRAPARAMS,
+                  payload: {
+                    extraParams: { fqn: provisionedID.result.fullyqualifiedname }
+                  },
+                });
+                return;
+              }
+            }
+          }
+        })
+      })
+    }
+  }, [idsloaded])
```
It seems that the fqn in extraParams is getting passed in through the processVerusId function so this is maybe no longer necessary? If so I'd prefer it to be deleted, as an automatic way to detect if a login-consent request is an interesting concept, but this does a few things incorrectly. Some of those include ignoring ID_FULLYQUALIFIEDNAME_VDXF_KEY as a subject type, which is a valid alternative to ID_ADDRESS_VDXF_KEY, and not accounting for potentially multiple subjects. If its necessary than we should discuss refactoring this.

michaeltout commented on src/reducers/deeplink.js:
```diff
@@ -1,13 +1,16 @@
-      id: null
+      id: null,
+      fromService: null,
+      extraParams: null
```
After reading through the code, I understand why this is necessary, seeing as data is reserved for data parsed from the deeplink itself, and in some cases we need data to be passed through the deeplink reducer, but could you please rename it to "passthrough" so its more clear what purpose it serves, and then renaming the "fqn" parameter you put here to automatically bring up the link identity to "fqnToAutolink" or something along those lines to also make it clear that putting data there will automatically prompt the user to link the id when the appropriate screen is reached.

michaeltout replied:
Also, now that you've added an extra field to the deeplink reducer, you need to edit resetDeeplinkData to clear out "extraParams" (or "passthrough" after the change) so that data isn't left there when another deeplink is scanned.

michaeltout commented on src/actions/actions/services/dispatchers/verusid/verusid.js:
```diff
@@ -191,10 +192,18 @@ export const checkVerusIdNotificationsForUpdates = async () => {
+            req.fromBuffer(Buffer.from(pendingIds[ticker][iaddress].loginRequest, 'base64'));
+            const verified = await verifyIdProvisioningResponse(system, response.data);
+
+            if (responseData.signing_id !== req.signing_id && !verified) {
```
Should be an "or" statement, otherwise it will take both cases of failing verification to actually fail.

michaeltout commented on src/containers/Home/HomeWidgets/NotificationWidget.js:
```diff
@@ -0,0 +1,283 @@
+    }
+}
+
+const getNotifications = ({ directory } = notifications, acchash) => {
```
I think notifications is undefined here, might cause a crash if you don't pass in a value for { directory } and it tries to default to notifications.

michaeltout commented on src/containers/DeepLink/LoginRequestInfo/LoginRequestInfo.js:
```diff
@@ -97,6 +102,74 @@ const LoginRequestInfo = props => {
+  useEffect(() => {
+
+    if (req && req.challenge && req.challenge.requested_access) {
+      if (req.challenge.requested_access.length === 1) {
```
If a login consent request only contained the primitives.IDENTITY_AGREEMENT requested_access type, it seems this would allow the user to continue through the request without being informed of what they need to agree to. As the identity agreement flow isn't part of the VerusID provisioning flow (as far as I know?), I would recommend removing ID agreement code from the PR for now and reintroducing it in a future PR.

michaeltout commented on src/containers/DeepLink/LoginRequestComplete/LoginRequestComplete.js:
```diff
@@ -70,6 +73,7 @@ const LoginRequestComplete = props => {
       if (fromService) {
         // Only return to services, default is main home screen.
         props.navigation.navigate("ServicesHome", { screen: fromService, params: { data: returnedData.data }});
+        dispatch(resetDeeplinkData());
```
Pull in latest changes from the multicurrency branch to fix this, it seems this isn't reflecting this commit https://github.com/VerusCoin/Verus-Mobile/commit/eba3b6250d86a7d42ada898454b77920ab42284c

michaeltout commented on src/containers/Home/HomeWidgets/NotificationWidget.js:
Looks good. Please run this file through a code formatter to get it to be formatted with 2 space indents so its consistent with the rest of the codebase.

michaeltout commented on src/actions/actions/services/dispatchers/verusid/verusid.js:
```diff
@@ -56,3 +70,253 @@ export const unlinkVerusId = async (iAddress, chain) => {
+export const checkVerusIdNotificationsForUpdates = async () => {
+  const state = store.getState();
+
+  const getPotentialPrimaryAddresses = async (coinObj, channel) => {
+    const seeds = await requestSeeds();
+
+    const seed = seeds[channel];
+
+    const keyObj = await deriveKeyPair(seed, coinObj, channel);
+    const {addresses} = keyObj;
+
+    return addresses;
+  };
+
+  if (state.authentication.activeAccount == null) {
+    throw new Error('You must be signed in for ID provisioning functions');
+  }
+  const isTestnet = Object.keys(state.authentication.activeAccount.testnetOverrides).length > 0;
```
This function relies on activeAccount not being null, so we can rely on activeAccount.keys existing. This means it would be preferable to get addresses from activeAccount.keys[coinObj][channel].addresses instead of accessing the seeds themselves.

michaeltout commented on src/containers/DeepLink/LoginRequestIdentity/LoginRequestIdentity.js:
```diff
@@ -39,7 +62,18 @@ const LoginRequestIdentity = props => {
-  const { system_id } = req
+  const { system_id } = req;
+
+  const getPotentialPrimaryAddresses = async () => {
```
This function seems unused and can be deleted?

michaeltout commented on src/components/SendModal/ProvisionIdentity/ProvisionIdentityResult/ProvisionIdentityResult.js:
```diff
@@ -2,17 +2,25 @@ import {Component} from 'react';
 
     this.state = {
-      response: props.route.params == null ? null : props.route.params.response,
+      response: new primitives.LoginConsentProvisioningResponse(props.route.params.response),
```
props.route.params can be null, which would cause this to crash
